cgen error on sizeof call with matching array type of sumtype

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 cfeb1bb, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 5ec9bb5.cfeb1bb
|:-------------------|:-------------------
|OS                  |linux, Linux version 6.6.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000
|Processor           |16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P
|Memory              |0.71GB/15.35GB
|                    |
|V executable        |/home/ge/.vlang/v
|V last modified time|2025-02-20 07:30:30
|                    |
|V home dir          |OK, value: /home/ge/.vlang
|VMODULES            |OK, value: /home/ge/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/ge/Code/v/zmq
|                    |
|Git version         |git version 2.48.1
|V git status        |weekly.2025.08-15-gcfeb1bb5
|.git/config present |true
|                    |
|cc version          |cc (GCC) 14.2.1 20250207
|gcc version         |gcc (GCC) 14.2.1 20250207
|clang version       |clang version 19.1.7
|tcc version         |tcc version 0.9.28rc 2024-07-31 HEAD@1cee0908 (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 0134e9b9
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.41

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  /tmp/match_err.vv && /tmp/match_err`
```v
module main

type Value = int | i64 | u64 | string | []u8

fn main() {
	val := Value(0)
	match val {
		int, i64, u64, []u8 {
			s := sizeof(val)
			println('Value is number or byte array, size=${s}')
		}
		string {
			println('Value is string')
		}
	}
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/match_err.01JMH483Y6AGNJT4YA1Z2M4M1D.tmp.c:5425: error: 'main__int___main__i64___main__u64___main__' undeclared
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

With `-show-c-output`:
```
======== Output of the C Compiler (/home/ge/.vlang/thirdparty/tcc/tcc.exe) ========
/tmp/v_1000/match_err.01JMH49E94N1X6Z3X2KDF76W9J.tmp.c:5425: error: 'main__int___main__i64___main__u64___main__' undeclared
===================================================================================
======== Output of the C Compiler (cc) ========
/tmp/v_1000/match_err.01JMH49E94N1X6Z3X2KDF76W9J.tmp.c: In function ‘main__main’:
/tmp/v_1000/match_err.01JMH49E94N1X6Z3X2KDF76W9J.tmp.c:5425:32: error: ‘main__int___main__i64___main__u64___main__’ undeclared (first use in this function)
 5425 |                 u32 s = sizeof(main__int___main__i64___main__u64___main__[]u8);
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_1000/match_err.01JMH49E94N1X6Z3X2KDF76W9J.tmp.c:5425:32: note: each undeclared identifier is reported only once for each function it appears in
/tmp/v_1000/match_err.01JMH49E94N1X6Z3X2KDF76W9J.tmp.c:5425:75: error: expected expression before ‘]’ token
 5425 |                 u32 s = sizeof(main__int___main__i64___main__u64___main__[]u8);
      |                                                                           ^
/tmp/v_1000/match_err.01JMH49E94N1X6Z3X2KDF76W9J.tmp.c:5429:75: error: expected expression before ‘]’ token
 5429 |                 u32 s = sizeof(main__int___main__i64___main__u64___main__[]u8);
      |                                                                           ^
/tmp/v_1000/match_err.01JMH49E94N1X6Z3X2KDF76W9J.tmp.c:5433:75: error: expected expression before ‘]’ token
 5433 |                 u32 s = sizeof(main__int___main__i64___main__u64___main__[]u8);
      |                                                                           ^
/tmp/v_1000/match_err.01JMH49E94N1X6Z3X2KDF76W9J.tmp.c:5437:75: error: expected expression before ‘]’ token
 5437 |                 u32 s = sizeof(main__int___main__i64___main__u64___main__[]u8);
      |                                                                           ^
===============================================
```

**What did you expect to see?**

No C errors
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/ast/ast.v
+++ b/vlib/v/ast/ast.v
@@ -884,6 +884,7 @@ pub enum ComptimeVarKind {
 	generic_param // generic fn parameter
 	generic_var   // generic var
 	smartcast     // smart cast when used in `is v` (when `v` is from $for .variants)
+	aggregate     // aggregate var
 }
 
 @[minify]

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -4331,6 +4331,8 @@ fn (mut c Checker) smartcast(mut expr ast.Expr, cur_type ast.Type, to_type_ ast.
 				is_inherited = expr.obj.is_inherited
 				ct_type_var = if is_comptime {
 					.smartcast
+				} else if c.table.type_kind(to_type_) == .aggregate {
+					.aggregate
 				} else {
 					.no_comptime
 				}

--- a/vlib/v/checker/str.v
+++ b/vlib/v/checker/str.v
@@ -32,7 +32,7 @@ fn (mut c Checker) get_default_fmt(ftyp ast.Type, typ ast.Type) u8 {
 			return `s`
 		}
 		if ftyp in [ast.string_type, ast.bool_type]
-			|| sym.kind in [.enum, .array, .array_fixed, .struct, .map, .multi_return, .sum_type, .interface, .none]
+			|| sym.kind in [.enum, .array, .array_fixed, .struct, .map, .multi_return, .sum_type, .interface, .aggregate, .none]
 			|| ftyp.has_option_or_result() || sym.has_method('str') {
 			return `s`
 		} else {

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -7306,13 +7306,7 @@ fn (mut g Gen) get_type(typ ast.Type) ast.Type {
 }
 
 fn (mut g Gen) size_of(node ast.SizeOf) {
-	typ_sym := g.table.sym(node.typ)
-	val_type := if typ_sym.info is ast.Aggregate {
-		typ_sym.info.types[g.aggregate_type_idx]
-	} else {
-		node.typ
-	}
-	typ := g.type_resolver.typeof_type(node.expr, g.get_type(val_type))
+	typ := g.type_resolver.typeof_type(node.expr, g.get_type(node.typ))
 	node_typ := g.unwrap_generic(typ)
 	sym := g.table.sym(node_typ)
 	if sym.language == .v && sym.kind in [.placeholder, .any] {

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5019,7 +5019,7 @@ fn (mut g Gen) ident(node ast.Ident) {
 	if node.info is ast.IdentVar {
 		if node.obj is ast.Var {
 			if !g.is_assign_lhs
-				&& node.obj.ct_type_var !in [.smartcast, .generic_param, .no_comptime] {
+				&& node.obj.ct_type_var !in [.smartcast, .generic_param, .no_comptime, .aggregate] {
 				comptime_type := g.type_resolver.get_type(node)
 				if comptime_type.has_flag(.option) {
 					if (g.inside_opt_or_res || g.left_is_opt) && node.or_expr.kind == .absent {
@@ -7306,7 +7306,13 @@ fn (mut g Gen) get_type(typ ast.Type) ast.Type {
 }
 
 fn (mut g Gen) size_of(node ast.SizeOf) {
-	typ := g.type_resolver.typeof_type(node.expr, g.get_type(node.typ))
+	typ_sym := g.table.sym(node.typ)
+	val_type := if typ_sym.info is ast.Aggregate {
+		typ_sym.info.types[g.aggregate_type_idx]
+	} else {
+		node.typ
+	}
+	typ := g.type_resolver.typeof_type(node.expr, g.get_type(val_type))
 	node_typ := g.unwrap_generic(typ)
 	sym := g.table.sym(node_typ)
 	if sym.language == .v && sym.kind in [.placeholder, .any] {

--- a/vlib/v/gen/c/infix.v
+++ b/vlib/v/gen/c/infix.v
@@ -836,11 +836,7 @@ fn (mut g Gen) infix_expr_in_optimization(left ast.Expr, left_type ast.Type, rig
 fn (mut g Gen) infix_expr_is_op(node ast.InfixExpr) {
 	mut left_sym := g.table.sym(g.unwrap_generic(g.type_resolver.get_type_or_default(node.left,
 		node.left_type)))
-	is_aggregate := left_sym.kind == .aggregate
-	if is_aggregate {
-		parent_left_type := (left_sym.info as ast.Aggregate).sum_type
-		left_sym = g.table.sym(parent_left_type)
-	}
+	is_aggregate := node.left is ast.Ident && g.comptime.get_ct_type_var(node.left) == .aggregate
 	right_sym := g.table.sym(node.right_type)
 	if left_sym.kind == .interface && right_sym.kind == .interface {
 		g.gen_interface_is_op(node)
@@ -880,7 +876,7 @@ fn (mut g Gen) infix_expr_is_op(node ast.InfixExpr) {
 		sub_sym := g.table.sym(sub_type)
 		g.write('_${left_sym.cname}_${sub_sym.cname}_index')
 		return
-	} else if left_sym.kind == .sum_type {
+	} else if left_sym.kind == .sum_type || is_aggregate {
 		g.write('_typ ${cmp_op} ')
 	}
 	if node.right is ast.None {

--- a/vlib/v/gen/c/match.v
+++ b/vlib/v/gen/c/match.v
@@ -211,6 +211,9 @@ fn (mut g Gen) match_expr_sumtype(node ast.MatchExpr, is_expr bool, cond_var str
 				g.write(cond_var)
 				cur_expr := unsafe { &branch.exprs[sumtype_index] }
 				if cond_sym.kind == .sum_type {
+					if cur_expr is ast.TypeNode {
+						g.type_resolver.update_ct_type(cond_var, cur_expr.typ)
+					}
 					g.write('${dot_or_ptr}_typ == ')
 					if cur_expr is ast.None {
 						g.write('${ast.none_type.idx()} /* none */')

--- a/vlib/v/tests/comptime/comptime_aggregate_var_test.v
+++ b/vlib/v/tests/comptime/comptime_aggregate_var_test.v
@@ -1,0 +1,25 @@
+module main
+
+type Value = int | i64 | u64 | string | []u8
+
+fn d(val Value) string {
+	match val {
+		int, i64, u64, []u8 {
+			s := sizeof(val)
+			x := val
+			return 'Value is number or byte array, size=${s} ${x}'
+		}
+		string {
+			x := val
+			return 'Value is string: ${x}'
+		}
+	}
+}
+
+fn test_main() {
+	assert d(Value(0)) == 'Value is number or byte array, size=4 0'
+	assert d(Value(i64(1))) == 'Value is number or byte array, size=8 1'
+	assert d(Value(u64(2))) == 'Value is number or byte array, size=8 2'
+	assert d(Value([u8(1), 2])) == 'Value is number or byte array, size=32 [1, 2]'
+	assert d(Value('')) == 'Value is string: '
+}

--- a/vlib/v/tests/typeof_aggregate_test.v
+++ b/vlib/v/tests/typeof_aggregate_test.v
@@ -41,6 +41,6 @@ fn test_typeof_aggregate() {
 
 	assert rets.len == 3
 	assert rets[0] == 'The type of `a` is `Foo`'
-	assert rets[1] == 'The type of `a` is `(Bar | Baz | Bazaar)` and its text says bar'
-	assert rets[2] == 'The type of `a` is `(Bar | Baz | Bazaar)` and its text says baz'
+	assert rets[1] == 'The type of `a` is `Bar` and its text says bar'
+	assert rets[2] == 'The type of `a` is `Baz` and its text says baz'
 }

--- a/vlib/v/type_resolver/type_resolver.v
+++ b/vlib/v/type_resolver/type_resolver.v
@@ -190,6 +190,13 @@ pub fn (mut t TypeResolver) get_type(node ast.Expr) ast.Type {
 						ctyp
 					}
 				}
+				.aggregate {
+					t.get_ct_type_or_default(node.name, if node.obj.smartcasts.len > 0 {
+						node.obj.smartcasts.last()
+					} else {
+						node.obj.typ
+					})
+				}
 				.key_var, .value_var {
 					// key and value variables from normal for stmt
 					t.get_ct_type_or_default(node.name, ast.void_type)

